### PR TITLE
 Fixes all initial calls to start() from fragments to viewmodels

### PR DESCRIPTION
**Pull request by JoseAlcerreca**
No duplicate loads in ViewModels anymore
Also fixes the ScrollChildSwipeRefreshLayout. Now it's correctly implemented in all 4 fragments.

### CODE REVIEW

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt [hunk 1]
```diff
@@ -54,7 +55,8 @@ class AddEditTaskFragment : Fragment() {
         super.onActivityCreated(savedInstanceState)
         setupSnackbar()
         setupNavigation()
-        loadData()
+        this.setupRefreshLayout(viewDataBinding.refreshLayout)
+        viewModel.start(getTaskId())
```
Comment (ceruleanotter) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: In `TaskDetailFragment ` you simply call the code that you've extracted here to getTaskId - I would remove the getTaskId, since I think this is the only place you call it.

Also I think you should be using by navArgs: https://developer.android.com/guide/navigation/navigation-pass-data

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt [hunk 2]
```diff
@@ -95,7 +89,13 @@ class TaskDetailFragment : Fragment() {
                 }
             }
         }
-        viewDataBinding.setLifecycleOwner(this.viewLifecycleOwner)
+        viewDataBinding.lifecycleOwner = this.viewLifecycleOwner
+
+        val taskId = arguments?.let {
+            TaskDetailFragmentArgs.fromBundle(it).TASKID
```
Comment (ceruleanotter) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt]: I think you should be using by navArgs: https://developer.android.com/guide/navigation/navigation-pass-data

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt [hunk 3]
```diff
@@ -95,7 +89,13 @@ class TaskDetailFragment : Fragment() {
...
+        }
+        viewDataBinding.viewmodel?.start(taskId)
```
Comment (ceruleanotter) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt]: just viewModel.start(taskId) (use the instance variable, not the viewmodel from the binding)

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt (hunk 1 above)
Comment (ceruleanotter) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: I still think start should be called in the initialization method of the viewmodel so that it isn't called again on a config change - is there a reason why it's being called from the creation methods?

File: app/src/test/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsUtilsTest.kt [hunk 4]
```diff
@@ -83,7 +81,6 @@ class StatisticsUtilsTest {
     }
 
     @Test
-    @Ignore
```
Comment (ceruleanotter) [inline app/src/test/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsUtilsTest.kt]: Seems a bit out of scope of the change?

File: app/src/main/res/layout/statistics_frag.xml [hunk 5]
```diff
@@ -17,63 +17,64 @@
...
 
         <variable
             name="stats"
-            type="com.example.android.architecture.blueprints.todoapp.statistics.StatisticsViewModel" />
+            type="com.example.android.architecture.blueprints.todoapp.statistics.StatisticsViewModel"/>
     </data>
 
-    <LinearLayout
+    <com.example.android.architecture.blueprints.todoapp.ScrollChildSwipeRefreshLayout
+        android:id="@+id/refresh_layout"
```
Comment (ceruleanotter) [inline app/src/main/res/layout/statistics_frag.xml:29]: nit: for a future change I want to standardize whether these ids are camelcase or _ --- what's the correct? There are examples of both in these layouts

**Review by ceruleanotter — CHANGES_REQUESTED** [review]
Added a few comments --- this is a big improvement!

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt (hunk 1 above)
Comment (JoseAlcerreca) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: Yes, we need the ID which is not available during init in VM. It would be complicated to pass it in the constructor so this is the best way.

Comment (JoseAlcerreca) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: Done

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt (hunk 2 above)
Comment (JoseAlcerreca) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt]: Done

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt (hunk 3 above)
Comment (JoseAlcerreca) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt]: Done

File: app/src/test/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsUtilsTest.kt (hunk 4 above)
Comment (JoseAlcerreca) [reply app/src/test/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsUtilsTest.kt]: Reverted

File: app/src/main/res/layout/statistics_frag.xml (hunk 5 above)
Comment (JoseAlcerreca) [reply app/src/main/res/layout/statistics_frag.xml:29]: Sounds good for a future change yes

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt (hunk 1 above)
Comment (ceruleanotter) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: Would it make sense to have some sort of check in start() where if the task ID hasn't changed, you don't reload?

Comment (JoseAlcerreca) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.kt]: Already has this

```
        if (isDataLoaded) {
            // No need to populate, already have data.
            return
        }
```

**Review by ceruleanotter — APPROVED** [review]
LGTM (though we still need to look into the fragment lifecycle stuff)
